### ISSUE TITLE
Selector support

### ISSUE THREAD

**Issue by beaver71** [rabbitmq/rabbitmq-server#2592]
I tried rhea node.js library and in particular this example https://github.com/amqp/rhea/tree/master/examples/selector example but obviously it's not working with RabbitMQ+plugin amqp1.0.
On the contrary it works with several other AMQP 1.0 brokers (ActiveMQ, Qpid cpp, Qpid Java, ...).
Do you plan to support filters, like JMS filters?

**Comment by stolsvik** [comment]
Does this also concern message selectors for queues when used with the JMS client? I rely on them to create consumers that specifically consume messages with high priority, which works well with Apache's ActiveMQ and Artemis, and it would be awesome if RabbitMQ also supported them.

**Comment by michaelklishin** [comment]
@beaver71 [4.0 release notes are public](https://github.com/rabbitmq/rabbitmq-server/releases/tag/v4.0.1).

**Comment by acogoluegnes** [comment]
4.1 will support [AMQP filter expressions](https://github.com/rabbitmq/rabbitmq-server/pull/12415) (for AMQP consumers consuming from streams).

The RabbitMQ JMS client supports JMS filters for topics.

Filters (whether AMQP or JMS) for queues are tricky to implement in a scalable way, but we are considering implementing them in the future.

**Comment by ansd** [comment]
Support for broker-side SQL filtering when consuming from streams using AMQP 1.0 is supported since RabbitMQ 4.2 and is described in https://www.rabbitmq.com/blog/2025/09/23/sql-filter-expressions .

Support for JMS message selectors (when consuming from a queue using AMQP 1.0) is currently under active development.

**Comment by beaver71** [comment]
> Support for broker-side SQL filtering when consuming from streams using AMQP 1.0 is supported since RabbitMQ 4.2 and is described in https://www.rabbitmq.com/blog/2025/09/23/sql-filter-expressions .

We've been interested in the SQL filters feature supported by RabbitMQ 4.2 streams. As described in the post, we've successfully tested the Java client https://github.com/rabbitmq/rabbitmq-amqp-java-client. Is it possible to use the same feature using a generic AMQP 1.0 client, such as Qpid Proton or Rhea.js?

 ---

> Support for JMS message selectors (when consuming from a queue using AMQP 1.0) is currently under active development.

Do you have a roadmap for this development?

**Comment by ansd** [comment]
> Is it possible to use the same feature using a generic AMQP 1.0 client, such as Qpid Proton or Rhea.js?

Yes, it is possible:
https://github.com/rabbitmq/rabbitmq-amqp-java-client uses the generic `protonj2` client underneath.
https://github.com/coders51/rabbitmq-amqp-js-client uses the generic `rhea` client underneath and also supports SQL filters via https://github.com/coders51/rabbitmq-amqp-js-client/pull/75 .

The SQL syntax follows AMQP extension specification https://docs.oasis-open.org/amqp/filtex/v1.0/csd01/filtex-v1.0-csd01.html So, the SQL syntax is not RabbitMQ specific.

> Do you have a roadmap for this development?

Yes, JMS message selectors (when consuming from a queue using AMQP 1.0) is on our roadmap for VMware Tanzu RabbitMQ.

**Comment by beaver71** [comment]
The specification of **AMQP filters** (https://docs.oasis-open.org/amqp/filtex/v1.0/filtex-v1.0.html) is currently in draft status, whereas the guidelines indicated on the official AMQP protocol website appear to be currently in force: https://www.amqp.org/specification/1.0/filters.

Therefore, reference is made to **Apache filters/JMS filters**, accessible via the following link: https://github.com/apache/qpid/blob/a7cacfb3e0cec373e6ef96e4f61439dfa0c36971/qpid/specs/apache-filters.xml. Indeed, some of the most widely used AMQP 1.0 brokers—specifically ActiveMQ Classic and ActiveMQ Artemis—adopt the aforementioned **JMS selectors** (https://artemis.apache.org/components/artemis/documentation/latest/filter-expressions.html), while they do not implement AMQP filters.

We also reference the international **C-Roads project**, co-funded by the EU (https://www.c-roads.eu/platform.html) and active in the ITS (Intelligent Transport Systems) sector, which has adopted the AMQP 1.0 standard as the interoperable interface between operators from different European nations. In more detail, the C-Roads specification requires the use of JMS selectors/filters as defined in the Apache context mentioned above.

So it would likely be beneficial for RabbitMQ to also offer compatibility with JMS filters, which are substantially equivalent to AMQP filters - Sql filter expressions (see https://docs.oasis-open.org/amqp/filtex/v1.0/csd01/filtex-v1.0-csd01.html#_Toc67929276).

This would involve considering the following type defined in AMQP filters:

```
<type name="sql-filter" class="restricted" source="string" provides="filter">
    <descriptor name="amqp:sql-filter" code="0x00000000:0x00000120"/>
</type>
```

as equivalent to the following:

```
<type name="selector-filter" class="restricted" provides="filter" source="string">
    <descriptor name="apache.org:selector-filter:string" code="0x0000468C:0x00000004"/>
</type>
```

or even to the following (as implemented, for example, in ActiveMQ Artemis and [used by the Rhea.js library](https://github.com/amqp/rhea/blob/main/lib/filter.js#L22C9-L22C78)):

```
<type name="jms-selector" class="restricted" provides="filter" source="string">
```

**Comment by ansd** [comment]
Correct, the SQL syntax of [AMQP SQL Filter Expressions](https://docs.oasis-open.org/amqp/filtex/v1.0/csd01/filtex-v1.0-csd01.html#_Toc67929276) as used in [Broker-Side SQL Filtering with RabbitMQ Streams](https://www.rabbitmq.com/blog/2025/09/23/sql-filter-expressions) is a **different** SQL syntax to the [JMS message selector syntax](https://jakarta.ee/specifications/messaging/3.1/jakarta-messaging-spec-3.1#message-selector-syntax).

> So it would likely be beneficial for RabbitMQ to also offer compatibility with JMS filters

Yes, let me repeat my answer from [above](https://github.com/rabbitmq/rabbitmq-server/issues/2592#issuecomment-3816916599): JMS message selectors (when consuming from a queue using AMQP 1.0) is on our roadmap for VMware Tanzu RabbitMQ and currently under active development.

**Comment by ansd** [comment]
Closing this issue since JMS message selectors will be supported in VMware Tanzu RabbitMQ 4.3
